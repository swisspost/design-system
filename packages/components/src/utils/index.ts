export * from './get-element-height';
export * from './on-transition-end';
export * from './property-checkers';
export * from './should-reduce-motion';
