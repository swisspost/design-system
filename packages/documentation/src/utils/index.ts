export * from './component-properties';
export * from './get-attributes';
export * from './map-classes';
export * from './spread-args';
