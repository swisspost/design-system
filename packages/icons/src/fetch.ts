import type { IconSet, SourceReport } from './models/icon.model';
import iconSets from './iconsets.config';
import { checkEnvVarsExist } from './utilities/environment';
import { setup } from './utilities/download/setup';
import { getBaseSourceReport, coloredLogMessage } from './utilities/shared';
import { fetchPage } from './utilities/download/fetchPage';
import { fetchFile } from './utilities/download/fetchFile';
import { format } from './utilities/download/format';
import { updateReport, writeReport } from './utilities/download/report';
import buildSVGs from './utilities/build';

async function fetchSVGs() {
  setup();

  for (const iconSet of iconSets) {
    console.log(coloredLogMessage(`<blue>Start downloading "${iconSet.name}" icons...</blue>`));
    const report = await downloadIconSet(iconSet, getBaseSourceReport());
    console.log(
      coloredLogMessage(
        `<blue>Downloading "${iconSet.name}" finished.</blue>\nDownloaded <green>${report.stats.success}</green> icons, <red>${report.stats.errors}</red> errored, <red>${report.stats.noSVG}</red> with wrong format (no SVG).\n`,
      ),
    );
  }

  buildSVGs();
}

async function downloadIconSet(
  iconSet: IconSet,
  report: SourceReport,
  nextPage?: string,
): Promise<SourceReport> {
  const body = await fetchPage(nextPage ?? iconSet.apiUrl);

  if (body === undefined) {
    throw new Error(`Fetch icons failed, response was ${body}`);
  }

  if ('error' in body) {
    throw new Error(`Fetch icons failed: ${body.error}`);
  }

  console.log(
    coloredLogMessage(
      `<yellow>Fetching icons</yellow> ${body.offset} - ${body.offset + body.count} of ${
        body['total-count']
      }`,
    ),
  );

  await Promise.all(
    format(body).map(async icon => {
      let svg: string | false = false;

      try {
        svg = await fetchFile(icon, iconSet.downloadDirectory);
      } catch (err) {
        console.log(err);
        icon.errored = true;

        if (err instanceof Error) {
          icon.errorMessage = err.message;
        }
      }

      report = updateReport(iconSet, svg, icon, report);
    }),
  );

  if (body.page.next) {
    // Recursively fetch more pages
    return downloadIconSet(iconSet, report, body.page.next);
  }

  return writeReport(iconSet, report);
}

if (checkEnvVarsExist()) {
  fetchSVGs();
}
