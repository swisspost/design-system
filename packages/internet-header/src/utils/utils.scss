@use '@swisspost/design-system-styles/components/fonts';
@use '@swisspost/design-system-styles/variables/color';
@use '@swisspost/design-system-styles/variables/breakpoints';
@use '@swisspost/design-system-styles/variables/type';
@use './mixins.scss';

*,
:host,
*::before,
*::after {
  box-sizing: border-box;
}

:host {
  font-family: type.$font-family-sans-serif;
}

button {
  font: inherit;
  padding: 0;
}

img,
svg {
  max-width: 100%;
  max-height: 100%;
}

svg {
  @media (forced-colors: active) {
    color: white;
  }
}

%no-list {
  list-style: none;
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
}

%blank-button {
  background-color: transparent;
  border: none;
  border-radius: 0;
  padding: 0;
}

%nav-link {
  text-decoration: none;
  color: rgba(color.$black, 0.8);
  transition: color 200ms;
  background-color: transparent;
  border-radius: 0;
  box-shadow: none;
  border: 0;
  margin: 0; // Buttons in safari have a default margin of 2px

  display: flex;
  align-items: center;
  gap: 0.5rem;

  &:hover,
  &:focus {
    color: black;
  }

  > svg {
    width: 1.4em;
    height: 1.4em;
    flex-shrink: 0;
  }

  > span {
    flex-shrink: 1;
  }
}

.box {
  > *:first-child {
    margin-top: 0;
  }

  > *:last-child {
    margin-bottom: 0;
  }
}

.mirrored {
  transform: scaleX(-1);
}

.rotate-270 {
  transform: rotate(270deg);
}

.bold {
  font-weight: 700;
}

.light {
  font-weight: 300;
}

.d-flex {
  display: flex;
}

.d-inline-flex {
  display: inline-flex;
}

.align-items-center {
  align-items: center;
}

.nav-link {
  @extend %nav-link;
}

.no-list {
  @extend %no-list;
}

.btn-blank {
  @extend %blank-button;
}

.wide-container {
  @media (min-width: 1441px) {
    margin: 0 auto;
    max-width: 1440px;
  }
}

.container {
  @include mixins.container-padding();
}

%visually-hidden {
  position: absolute;
  width: 1px;
  height: 1px;
  border: 0;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0 0 0 0);
}

.visually-hidden {
  @extend %visually-hidden;
}

@each $breakpoint, $value in breakpoints.$grid-breakpoints {
  .hidden-#{$breakpoint} {
    @include mixins.max($breakpoint) {
      display: none;
    }
  }
}
