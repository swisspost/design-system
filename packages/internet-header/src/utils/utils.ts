/**
 * Generates a (sequential) unique ID. If prefix is given, the ID is appended to it.
 *
 * @param prefix The value to prefix the ID with
 * @returns Unique ID
 */
export const uniqueId = (() => {
  let num = 0;
  return function (prefix?: string) {
    prefix = prefix || '';
    num += 1;
    return prefix + num;
  };
})();

export const prefersReducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches;
