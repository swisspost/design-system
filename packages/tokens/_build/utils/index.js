export { default as objectDeepmerge } from './object-deepmerge.js';
export { default as objectTextoutput } from './object-textoutput.js';
