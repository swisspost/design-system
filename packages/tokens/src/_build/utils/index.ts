export { default as objectDeepmerge } from './object-deepmerge.ts';
export { default as objectTextoutput } from './object-textoutput.ts';
