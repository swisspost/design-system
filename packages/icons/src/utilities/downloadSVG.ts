import fetch from 'node-fetch';
import fs from 'fs';
import path from 'path';
import { optimize } from 'svgo';
import svgoOptions from '../../svgo.config';
import { IIcon } from '../models/icon.model';
import { getRequestInit } from './environment';

// Attempt to sanitize the received svg string
const extractSVG = (input: string) => {
  const regex = new RegExp(/<svg[\S\s]*<\/svg>/m);
  const matches = input.match(regex);
  return matches === null ? input : matches[0];
};

export const downloadSVG = async (icon: IIcon, output: string) => {
  if (!icon.meta.downloadLink) {
    return false;
  }

  try {
    const svg = await fetch(icon.meta.downloadLink, getRequestInit());
    const svgString = await svg.text();
    const optimizedSvg = optimize(extractSVG(svgString), svgoOptions);
    const optimizedName = icon.file.name
      .split('.')
      .map(token => token.trim())
      .join('.');

    fs.writeFileSync(path.join(output, optimizedName), optimizedSvg.data);
    return optimizedSvg.data;
  } catch (err) {
    console.log(`SVG Download error: ${err} @ ${icon.meta.downloadLink}`);
    throw err;
  }
};
