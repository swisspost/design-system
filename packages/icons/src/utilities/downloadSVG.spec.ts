import { mocked } from 'ts-jest/utils';
import fetch, { Response } from 'node-fetch';
import { Businessfield, Type, TypeFilter, VariantMIME } from '../models/censhare-result-page.model';
import mockFs from 'mock-fs';
import { downloadSVG } from './downloadSVG';

jest.mock('node-fetch');

describe('downloadSVG', () => {
  it('should download an svg icon', async () => {
    const outputPath = './icons';
    mockFs({
      [outputPath]: {},
    });

    mocked(fetch).mockImplementationOnce(() =>
      Promise.resolve({
        text: () => Promise.resolve('<svg><path d="M16 8.4l-12.533 12.4z"></path></svg>'),
      } as Response),
    );

    const svg = await downloadSVG(
      {
        uuid: '00000000-0000-0000-0000-000000000000',
        id: 0,
        type: Type.PicturePictogram,
        typeFilter: TypeFilter.Pictograms,
        meta: {
          downloadLink: '/test',
          businessfield: Businessfield.Kommunikation,
          keywords: [
            'Test',
            'Test2',
          ],
          year: ''
        },
        file: {
          mime: VariantMIME.ImageSVGXML,
          name: 'test.svg',
          basename: 'test',
          ext: '.svg',
          size: {
            width: 32,
            dpi: 72,
            height: 32
          }
        },
        createdAt: new Date(),
        modifiedAt: new Date()
      },
      outputPath,
    );

    mockFs.restore();

    expect(svg).toMatchSnapshot();
  });
});
