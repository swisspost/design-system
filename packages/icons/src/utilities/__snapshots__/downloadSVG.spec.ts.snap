// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`downloadSVG should download an svg icon 1`] = `
"<svg>
  <path d="M16 8.4 3.467 20.8z"/>
</svg>
"
`;
